tests/TestCompiler.hs

```diff
@@ -84,41 +81,36 @@ warningTests
   :: [P.Module]
   -> [P.ExternsFile]
   -> M.Map P.ModuleName FilePath
-  -> IO TestTree
+  -> Spec
 warningTests supportModules supportExterns supportForeigns = do
-  warningTestCases <- getTestFiles "warning"
-  tests <- forM warningTestCases $ \testPurs -> do
-    let mainPath = getTestMain testPurs
-    expectedWarnings <- getShouldWarnWith mainPath
-    wTc <- testSpecs $
-      it ("'" <> takeFileName mainPath <> "' should compile with warning(s) '" <> intercalate "', '" expectedWarnings <> "'") $
+  warningTestCases <- runIO $ getTestFiles "warning"
+
+  describe "Warning examples" $
+    forM_ warningTestCases $ \testPurs -> do
+      let mainPath = getTestMain testPurs
+      expectedWarnings <- runIO $ getShouldWarnWith mainPath
+      it ("'" <> takeFileName mainPath <> "' should compile with warning(s) '" <> intercalate "', '" expectedWarnings <> "'") $ do
         assertCompilesWithWarnings supportModules supportExterns supportForeigns testPurs expectedWarnings
-    return $ wTc ++ [ goldenVsString
-                      ("'" <> takeFileName mainPath <> "' golden test")
-                      (replaceExtension mainPath ".out")
-                      (BS.fromStrict . T.encodeUtf8 . T.pack <$> printErrorOrWarning supportModules supportExterns supportForeigns testPurs)
-                    ]
-  return $ testGroup "Warning examples" $ concat tests
+        goldenVsString
```

Comment:
It occurred to me while I was doing this that we are actually compiling everything in here twice: once when verifying the `shouldWarnWith`, and again when verifying the golden file output. That's also something I'd like to address in the future.